lib/active_merchant/billing/gateways/visanet_peru.rb

```diff
@@ -0,0 +1,239 @@
+        params = {}
+        action, merchant_id, purchase_number = split_authorization(authorization)
+        options[:merchant_id] = merchant_id
+        options[:purchaseNumber] = purchase_number
+        params[:externalTransactionId] = purchase_number
+        puts options
+        case action
+        when "authorize"
+          commit("void", params, options)
+        when "capture"
+          commit("cancel", params, options)
+        end
+      end
+
+      # def refund(amount, authorization, options={})
+      # end
+
+      # def credit(amount, payment_method, options={})
+      # end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      # def store(payment_method, options = {})
+      #   post = {}
+      #   add_payment_method(post, payment_method)
+      #   add_customer_data(post, options)
+
+      #   commit("store", post)
+      # end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        #JSON.
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r((\"cardNumber\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cvv2Code\\\":)\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      CURRENCY_CODES = Hash.new{|h,k| raise ArgumentError.new("Unsupported currency: #{k}")}
+      CURRENCY_CODES["USD"] = 840
+      CURRENCY_CODES["PEN"] = 604
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money).to_f
+        # Visanet Peru expects a 9-digit numeric purchaseNumber
+        post[:purchaseNumber] = options[:order_id]
+        post[:externalTransactionId] = options[:order_id]
+        post[:currencyId] = CURRENCY_CODES[options[:currency] || currency(money)]
+      end
+
+      def add_payment_method(post, payment_method)
+        post[:firstName] = payment_method.first_name
+        post[:lastName] = payment_method.last_name
+        post[:cardNumber] = payment_method.number
+        post[:cvv2Code] = Integer(payment_method.verification_value, 10)
+        post[:expirationYear] = format(payment_method.year, :four_digits)
+        post[:expirationMonth] = format(payment_method.month, :two_digits)
+      end
+
+      def add_customer_data(post, options)
+        post[:email] = options[:email]
+        antifraud = {}
+        billing_address = options[:billing_address] || options[:address]
+        if (billing_address)
+          antifraud[:billTo_street1] = billing_address[:address1]
+          antifraud[:billTo_city] = billing_address[:city]
+          antifraud[:billTo_state] = billing_address[:state]
+          antifraud[:billTo_country] = billing_address[:country]
+          antifraud[:billTo_postalCode]    = billing_address[:zip]
+        end
+        post[:antifraud] = antifraud
```

Comment:
Instead of initializing an empty `:antifraud` element here, perhaps we should check/initialize where it's needed instead, which I believe is just in `add_device_fingerprint_data` and `add_merchant_define_data`?

And while I'm looking at it, it seems those two methods aren't ever called independent of each other. Can we combine them into one `add_antifraud_data` method?